### ISSUE TITLE
Add support for generic flamegraph visualization type

### ISSUE THREAD

**Issue by Rperry2174** [grafana/grafana#53723]
## Why is a flamegraph visualization needed in Grafana :

First of all, huge fan of Grafana and how much its evolved since I started using it! However, one thing that it's currently missing that would fit in great (IMO) is more support around profiling / flamegraphs!

As profiling becomes a more popular visualization type it would be great if Grafana supported visualizing flamegraphs in both the explore tab and elsewhere across grafana. Flamegraphs could be used to represent various kinds of hierarchical data, but probably most commonly profiles and traces.

Unfortunately as it stands, there's plenty of ways to grab profiling data from somewhere, but there's very few ways to visualize it alongside your other observability data such as logs, traces, and metrics. The motivation for this issue came from a user asking for exactly this use case in the [Pyroscope repo](https://github.com/pyroscope-io/pyroscope/issues/1381#issuecomment-1211499454). 

### Where could a flamegraph visualization type be used?:
This could be used in various places:
- explore tab (for visualizing profiles)
- tracing tab (for representing traces as flamegraphs -- like [here in jaeger](https://github.com/jaegertracing/jaeger-ui/pull/976))
- regular panels (for visualizing data that makes sense as a flamegraph)

#### Flamegraph visualization in the explore tab
Here's an example of POC that links directly from Loki logs/traces to a corresponding profile represented as a flamegraph:

![flamegraphs_in_grafana](https://user-images.githubusercontent.com/23323466/184584085-744014dc-b043-4837-af07-2ceb61f81068.gif)
See source [here](https://github.com/pyroscope-io/pyroscope/tree/main/examples/tracing/jaeger)

#### Flamegraph visualization as a dashboard item
And an what it might look like to have a native profiling visualization type selectable from the sidebar in a regular dashboard panel:
![image](https://user-images.githubusercontent.com/23323466/184582143-e54a82f8-c18e-4434-9290-023a7c0b4d31.png)

While this profile is being stored in Pyroscope, examples of places where people generate profiles and view them somewhere other than Grafana -- but where it may be convenient to view inside of grafana:

- go (pprof profiles) - could be retrieved from pprof endpoint
- Java (async-profiler)
- perf (perf reports)
- eBPF
- Jaeger traces
- etc

Some examples of open source projects who are building observability tooling around profiling and would benefit from being able to visualize profiles in Grafana:

- [Pyroscope](https://github.com/pyroscope-io/pyroscope)
- [Parca](https://github.com/parca-dev/parca)
- [Pixie](https://github.com/pixie-io/pixie)
- [Profefe](https://github.com/profefe/profefe)
- [bloomberg / memray](https://github.com/bloomberg/memray)
- [janestreet / magic-trace](https://github.com/janestreet/magic-trace)

And there are many closed source ones I assume as well

## Potential Implementation Details
As far as implementation goes,it would be nice to have a generic flamegraph visualization type which takes one of the popular flamegraph formats as input and outputs a visualized flamegraph:

- [collapsed / folded](https://www.brendangregg.com/blog/2016-04-30/linux-perf-folded.html) (collapsed is the most common denominator between all profiling formats)
- [pprof](https://github.com/google/pprof/blob/main/proto/profile.proto)
- Potentially an [OTEL supported format](https://github.com/open-telemetry/oteps/issues/139)

Currently the [`preferredVisualizationTypes`](https://github.com/grafana/grafana/blob/main/packages/grafana-data/src/types/data.ts#L24-L25) only includes graph, table, logs, traces, and nodeGraph. And as a result, these are the only visualization types that are available within the explore tab. I think "profiles" fit in categorically with these other observability signal visualizations and could be available alongside them.

For the POC in the screenshot above we used Pyroscope's flamegraph component a generic [standalone NPM package](https://www.npmjs.com/package/@pyroscope/flamegraph) which can support several supported formats and render a flamegraph accordingly which may or may not be a convenient way to implement this feature easily.

Not sure what would justify adding a flamegraph as a generic visualization type (or what criteria it would need to fit to show up alongside graph, table,logs, traces, and nodegraph) so was hoping someone could answer a few questions:

1. Is there any plans for grafana to support flamegraphs as a visualization type?
2. If not is there any specific criteria that would need to be met for you to consider accepting a (generic) flamegraph visualization in grafana so that it can be shown in the explore tab?
3. If there's no plans for a generic flamegraph visualization type, is there any way to allow custom plugins in the explore tab (without forking grafana)?

## Implementation checklist
___

- [x] Is/could it be used in more than one place in Grafana?
- [x] Post screenshots possible.
- [ ] It has a single use case.
- [x] It is/could be used in multiple places.

- [ ] Component has a story in Storybook.
- [ ] Props and naming follows [our style guide](https://github.com/grafana/grafana/blob/main/contribute/style-guides/frontend.md).
- [ ] It is extendable (rest props are spread, styles with className work, and so on).
- [ ] Uses [theme for spacing, colors, and so on](https://github.com/grafana/grafana/blob/main/contribute/style-guides/themes.md).
- [ ] Works with both light and dark theme.

## Documentation

- [ ] Properties are documented.
- [ ] Use cases are described.
- [ ] Code examples for the different use cases.
- [ ] Dos and don'ts.
- [ ] Styling guidelines, specific color usage (if applicable).

**Comment by Rperry2174** [comment]
Adding that there is an effort in progress to support custom plugins in the explore tab:
https://github.com/grafana/grafana/pull/52057

**Comment by nserrino** [comment]
I work on the Pixie datasource plugin for Grafana and would love to see this - when crafting premade Grafana dashboards using the Pixie plugin, I was unable to include any flamegraphs. We considered investing in a custom flamegraph plugin for Pixie, but it would be great to see something more standardized.

**Comment by Rperry2174** [comment]
@nserrino what format does the pixie datasource plugin produce (or what formats is it capable of returning to the frontend)? i.e. pprof, collapsed, etc

**Comment by nserrino** [comment]
@Rperry2174 With some effort, Pixie can support either pprof or collapsed, and collapsed is probably easier. Depending on what standard the generic flamegraph visualization wants to rely on, it could go either way, so it will be good for me to stay in the loop on what's decided here so we can prioritize accordingly.

**Comment by aocenas** [comment]
@Rperry2174 we recently added a flamegraph implementation behind a feature flag https://github.com/grafana/grafana/pull/56376. Sorry for this being a bit secret until now but we tried to quickly build and iterate something usable. There isn't much documentation for it as it is behind feature flag it's not a GA feature but hope that will come soon.

That said we did not abandon our effort of allowing custom visualizations in Explore https://github.com/grafana/grafana/pull/52057. That work just need more time and thought and we didn't want to be blocked by it.

Note regarding the data format. In Grafana we are heavily trying to standardize on DataFrame abstraction which allows us to do things like transforms and expressions. So we cannot easily support any kind of format and as DataFrame is inherently a table like structure fitting a tree structure may seem a bit awkward. Right now we use a https://en.wikipedia.org/wiki/Nested_set_model where you just need a `level` field that encodes depth in the tree and by ordering items in depth first traversal order we can recreate the tree back from table data.

We will provide more documentation for it later on but if somebody wants more info about it just ping me.

**Comment by Rperry2174** [comment]
Thanks for the update @aocenas !!! Very excited for this -- interested to see a proof of concept of what to do with popular formats like collapsed and pprof and how those will be visualized using this plugin but seems like its almost there :)
